<a href="https://colab.research.google.com/github/kelseykyla/LinearRegression/blob/main/Linear_Regression_Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **SPLITTING THE DATASET INTO 50:50**

In [ ]:
#import libraries
import pandas as pd
import numpy as nm

In [ ]:
#Importing the dataset
nyc=pd.read_csv('/content/ave_hi_nyc_jan_1895-2018.csv.xls')
nyc

Date  Value  Anomaly
0    189501   34.2     -3.2
1    189601   34.7     -2.7
2    189701   35.5     -1.9
3    189801   39.6      2.2
4    189901   36.4     -1.0
..      ...    ...      ...
119  201401   35.5     -1.9
120  201501   36.1     -1.3
121  201601   40.8      3.4
122  201701   42.8      5.4
123  201801   38.7      1.3

[124 rows x 3 columns]

In [ ]:
nyc.columns=['Date','Temperature','Anomaly']
nyc

Date  Temperature  Anomaly
0    189501         34.2     -3.2
1    189601         34.7     -2.7
2    189701         35.5     -1.9
3    189801         39.6      2.2
4    189901         36.4     -1.0
..      ...          ...      ...
119  201401         35.5     -1.9
120  201501         36.1     -1.3
121  201601         40.8      3.4
122  201701         42.8      5.4
123  201801         38.7      1.3

[124 rows x 3 columns]

In [ ]:
nyc.Temperature.values

array([34.2, 34.7, 35.5, 39.6, 36.4, 37.4, 37. , 35. , 35.5, 29.8, 33.7,
       42.3, 40.5, 38.3, 39.6, 36.1, 40.7, 29.3, 46.9, 36.9, 38.9, 42.3,
       38.4, 26.5, 40.6, 29.7, 39.7, 33.4, 35.2, 39.4, 33.8, 37.2, 36.8,
       40. , 37.8, 38.5, 38.2, 47.6, 45.6, 39.8, 35. , 34.3, 45.7, 37.3,
       36.8, 30. , 34.5, 36.5, 36.1, 39. , 31.5, 40.2, 43. , 30.1, 43.5,
       47.6, 42.5, 42.4, 42.4, 37. , 35.6, 36.2, 33.8, 36.4, 37.1, 37.8,
       31.9, 38. , 34.6, 40.2, 34. , 37.1, 41.6, 31.7, 35.7, 29.9, 31.5,
       40.5, 40.6, 39.7, 41.7, 32.6, 26.1, 32.3, 38.6, 38. , 30.6, 30.9,
       38. , 33.5, 33. , 39.3, 35.9, 34.2, 42.4, 46. , 39.6, 40.9, 41. ,
       31.6, 42.6, 37. , 39.2, 45.8, 40.8, 37.8, 38.8, 45.4, 32.9, 31. ,
       37.4, 47.6, 43.6, 42.4, 33.4, 38.1, 34.8, 44.2, 40.8, 35.5, 36.1,
       40.8, 42.8, 38.7])

In [ ]:
#Checking for missing data
missing_data=nyc.isna().sum()
missing_data

Date           0
Temperature    0
Anomaly        0
dtype: int64

In [94]:
nyc.Date.values

array([189501, 189601, 189701, 189801, 189901, 190001, 190101, 190201,
       190301, 190401, 190501, 190601, 190701, 190801, 190901, 191001,
       191101, 191201, 191301, 191401, 191501, 191601, 191701, 191801,
       191901, 192001, 192101, 192201, 192301, 192401, 192501, 192601,
       192701, 192801, 192901, 193001, 193101, 193201, 193301, 193401,
       193501, 193601, 193701, 193801, 193901, 194001, 194101, 194201,
       194301, 194401, 194501, 194601, 194701, 194801, 194901, 195001,
       195101, 195201, 195301, 195401, 195501, 195601, 195701, 195801,
       195901, 196001, 196101, 196201, 196301, 196401, 196501, 196601,
       196701, 196801, 196901, 197001, 197101, 197201, 197301, 197401,
       197501, 197601, 197701, 197801, 197901, 198001, 198101, 198201,
       198301, 198401, 198501, 198601, 198701, 198801, 198901, 199001,
       199101, 199201, 199301, 199401, 199501, 199601, 199701, 199801,
       199901, 200001, 200101, 200201, 200301, 200401, 200501, 200601,
      

In [135]:
#Splitting the data into training and test set
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

#Reshaping the input data
X=nyc.Date.values.reshape(-1,1)
y=nyc.Temperature.values

#Splitting the data into training and test set
#Testing using 50:50
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=0)

In [114]:
#Feature Scaling of the dataset
from sklearn.preprocessing import StandardScaler
st_x = StandardScaler()
X_train = st_x.fit_transform(X_train)
X_test = st_x.transform(X_test)

In [123]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

#Training the model
model= linear_model.LinearRegression()
model.fit(X=X_train,y=y_train)

LinearRegression()

In [124]:
#Making predictions
y_predict=model.predict(X_test)

In [127]:
#Calculate RMSE
mse = mean_squared_error(y_test,y_predict)
rmse = np.sqrt(mse)
print(f'Root Mean Squared Error (RMSE):{rmse}')

Root Mean Squared Error (RMSE):4.362802610798153


In [126]:
#Displaying a few predictions
results = pd.DataFrame({'Actual': y_test, 'Predicted': y_predict})
print(results)

    Actual  Predicted
0     35.5  36.911078
1     31.5  37.784060
2     37.3  37.638563
3     40.6  37.243643
4     36.4  38.054269
..     ...        ...
57    45.7  37.617778
58    34.3  37.596992
59    36.4  36.827937
60    36.1  37.056575
61    29.3  37.098146

[62 rows x 2 columns]




**COMMENTS ON THE RMSE**

An RMSE of 4.36 is quite large. This suggests that the model's predictions are not very accurate, further improvement is needed

# **SPLITTING THE DATASET INTO 80:20**

In [159]:
#Split the data into training and testing sets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [160]:
#copy the original dataset
X_copy = X.copy()
y_copy = y.copy()

In [161]:
#Split the dataset into 80% and 20%
X_copy_train,X_copy_test,y_copy_train,y_copy_test = train_test_split(X_copy,y_copy,test_size=0.2,random_state=0)

In [162]:
#Initializing and Training the model
model = linear_model.LinearRegression()
model.fit(X_copy_train,y_copy_train)

LinearRegression()

In [163]:
print('Train Score: ', model.score(X_copy_train, y_copy_train))
print('Test Score: ', model.score(X_copy_test, y_copy_test))

Train Score:  0.010910532267929485
Test Score:  0.002894595508108222


In [164]:
#Making predictions
y_predict = model.predict(X_copy_test)

In [165]:
#Evaluating the model
mse = mean_squared_error(y_copy_test,y_predict)
rmse = np.sqrt(mse)
print(f'Root Mean Square: {rmse}')

Root Mean Square: 4.221927404224914


COMMENTS ON THE RMSE

The RMSE for the 80:20 split is slightly lower than the RMSE for the 50:50 split.
Both RMSE values are relatively close, indicating consistency in your model's performance across different splits

The 50:50 and 80:20RMSE values are quite similar, indicating that the model is performing consistently. The slightly better performance with the 80:20 split suggests that having more training data is beneficial, but the difference is not substantial enough to warrant major concerns.

# **SPLITTING THE DATASET INTO 90:10**

In [166]:
#Split the data into training and testing sets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

#copy the original dataset
X_copy = X.copy()
y_copy = y.copy()

#Split the dataset into 90% and 10%
X_copy_train,X_copy_test,y_copy_train,y_copy_test = train_test_split(X_copy,y_copy,test_size=0.1,random_state=0)

#Initializing and Training the model
model = linear_model.LinearRegression()
model.fit(X_copy_train,y_copy_train)


LinearRegression()

In [167]:
#Making predictions
y_predict = model.predict(X_copy_test)


In [168]:
#Evaluating the model
mse = mean_squared_error(y_copy_test,y_predict)
rmse = np.sqrt(mse)
print(f'Root Mean Square: {rmse}')

Root Mean Square: 4.261210279281796


**COMMENT ON THE RMSE**

The RMSE for the 90:10 split (4.26) is slightly higher than for the 80:20 split (4.22) but lower than the 50:50 split (4.36).
The differences among the RMSE values are relatively small, indicating consistent performance across different splits.


# **SPLITTING THE DATASET INTO 75:25**

In [173]:
#Split the data into training and testing sets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

#copy the original dataset
X_copy = X.copy()
y_copy = y.copy()

#Split the dataset into 90% and 10%
X_copy_train,X_copy_test,y_copy_train,y_copy_test = train_test_split(X_copy,y_copy,test_size=0.25,random_state=0)

#Initializing and Training the model
model = linear_model.LinearRegression()
model.fit(X_copy_train,y_copy_train)

LinearRegression()

In [174]:
#Making predictions
y_predict = model.predict(X_copy_test)

In [175]:
#Evaluating the model
mse = mean_squared_error(y_copy_test,y_predict)
rmse = np.sqrt(mse)
print(f'Root Mean Square: {rmse}')

Root Mean Square: 4.283797052351777


**COMMENT ON THE RMSE**

The RMSE for the 75:25 split (4.28) is close to the RMSE for the 90:10 split (4.26) and the 80:20 split (4.22), and slightly lower than the 50:50 split (4.36).
The differences among the RMSE values are small, indicating consistent model performance across various splits.

# **CONCLUSION**
80:20 Split: RMSE = 4.22

50:50 Split: RMSE = 4.36

90:10 Split: RMSE = 4.26

75:25 Split: RMSE = 4.28

**Comparison and Analysis:**
80:20 Split (RMSE = 4.22):

Provides a good balance between training and testing data.
Slightly lower RMSE indicates better performance compared to other splits.
50:50 Split (RMSE = 4.36):

Higher RMSE compared to other splits.
Indicates that having less training data might impact the model's performance.
90:10 Split (RMSE = 4.26):

Slightly higher RMSE than the 80:20 split.
Having a smaller test set might not provide a comprehensive evaluation, leading to a slight increase in RMSE.
75:25 Split (RMSE = 4.28):

RMSE is close to the 90:10 split.
Provides a good amount of training data while also having a sizable test set.
**Final Conclusion:**

The RMSE values for the 80:20, 90:10, and 75:25 splits are quite similar, with the 80:20 split having the lowest RMSE of 4.22. This suggests that the model performs best with the 80:20 split. The higher RMSE for the 50:50 split indicates that having more training data is beneficial for the model's performance.

**Recommendation:**

Based on the RMSE values, the 80:20 split is the best ratio to use. It provides a good balance between training and testing data, resulting in the lowest RMSE and indicating the best model performance among the splits analyzed.